spec.html

```diff
@@ -5783,6 +5783,29 @@ <h1>
           1. Return _integer_.
       </emu-alg>
     </emu-clause>
+
+    <emu-clause id="sec-toclampedindex" type="abstract operation">
+      <h1>
+        ToClampedIndex (
+          _value_: an ECMAScript language value,
+          _length_: a non-negative integer,
+          optional _lowerBound_: an integer,
+          optional _upperBound_: an integer,
+        ): either a normal completion containing an integer or a throw completion
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>It converts _value_ to an integer (interpreting negative values relative to _length_) and clamps the result to the inclusive interval from _lowerBound_ to _upperBound_ (respectively defaulting to 0 and _length_).</dd>
+      </dl>
+      <emu-alg>
+        1. If _lowerBound_ is not present, set _lowerBound_ to 0.
+        1. If _upperBound_ is not present, set _upperBound_ to _length_.
+        1. Assert: _lowerBound_ &le; _upperBound_.
+        1. Let _integer_ be ? ToIntegerOrInfinity(_value_).
+        1. If _integer_ is finite and _integer_ &lt; 0, set _integer_ to _length_ + _integer_.
+        1. Return the result of clamping _integer_ between _lowerBound_ and _upperBound_.
```

Comment:
> E.g., a caller that needs all those transforms might say:
> 
> ```
> Let _index_ be ? ToIntegerOrInfinity(_arg_).
> Set _index_ to HandleNegativeIndex(_index_, _len_).
> Set _index_ to ClampBelow(_index_, 0).
> Set _index_ to ClampAbove(_index_, _len_ - 1).
> ```

My goal in introducing ToClampedIndex is to eliminate that kind of noisy boilerplate so the relevant aspects of operations can be more visible, and I think it becomes a sufficiently beneficial pattern to justify use in `at` etc. even though they don't need clamping. This is definitely a matter of taste, though.

Reply:
I second @jmdyck's comment about this doing too many things. There's probably a factoring which can reduce some of the boilerplate, but I don't think it's a good idea to try to combine all this stuff into a single operation that can be used everywhere. Boilerplate is much better than having operations which are hard to follow.

I think the HandleNegativeIndex suggestion might be viable. Taking the example of `Array.prototype.at`, for example, right now it's

```
1. Let _relativeIndex_ be ? ToIntegerOrInfinity(_index_).
1. If _relativeIndex_ &ge; 0, then
  1. Let _k_ be _relativeIndex_.
1. Else,
  1. Let _k_ be _len_ + _relativeIndex_.
1. If _k_ &lt; 0 or _k_ &ge; _len_, return *undefined*.
```

With the current PR that becomes
```
1. Let _k_ be ? ToClampedIndex(_index_, _len_, -1).
1. If _k_ &lt; 0 or _k_ &ge; _len_, return *undefined*.
```
and you have to read quite carefully to understand that the range for the "clamped index" actually starts at _negative one_, and therefore will hit the `k < 0` case in the second line. With something like HandleNegativeIndex it could instead be
```
1. Let _relativeIndex_ be ? ToIntegerOrInfinity(_index_).
1. Let _k_ be HandleNegativeIndex(_relativeIndex_, _len_). 
1. If _k_ &lt; 0 or _k_ &ge; _len_, return *undefined*.
```
which is, I think, much easier to follow than trying to figure out what's going on with ToClampedIndex.

I could see a case for also having an operation like `ToClampedIndex` which does not take the final two optional parameters, for use in cases where it's actually doing the thing its name implies; that's a large majority of the callsites.

Reply:
Sure, I can see adding that for use in `at`s. I do like the extended-parameter ToClampedIndex for `lastIndexOf`s (where it supports "let _k_ be ? ToClampedIndex(_fromIndex_, _len_, -1, _len_ - 1)" rather than e.g. "let _k_ be min(ToAbsoluteIndex(_fromIndex_, _len_), _len_ - 1)"), but that's only two sites. What are your thoughts on those?

Reply:
Since it's only two callsites I'm inclined not to both trying to factor that logic out, and letting ToClampedIndex be simpler.

Reply:
Okay, done.